File: Project3/JavaScript/Task2.js
/* Project3 Task2
 * Tracy Quan
 *
 *
 * JavaScript
 * Task2: Write an example program in each language that executes a binary search
 * on a list or array of numbers.
 */

// return the index of the target item if found, return -1 if not found.
function binary_search(list, item){
  var min_index = 0;
  var max_index = list.length - 1;
  var guess_index;

  while (min_index <= max_index){
    guess_index = Math.floor((min_index + max_index)/2);
    if (list[guess_index] == item){
      return guess_index;
    }
    else if (list[guess_index] > item) {
      max_index = guess_index - 1;
    }
    else{
      min_index = guess_index + 1;
    }
  }

  return -1;

}

alert("The Index of 8 is " + binary_search([2,4,8,16,32,64,128,256,512,1024,2048],8));
alert("The Index of 1024 is " + binary_search([2,4,8,16,32,64,128,256,512,1024,2048],1024));

File: Project3/C++/Task2.cpp
/* Project3 Task2
 * Tracy Quan
 *
 *
 * C++
 * Task2:Write an example program in each language that executes a binary search
 * on a list or array of numbers.
 */

#include<stdio.h>

int binarySearch(int arr[], int first_index, int last_index, int target){
  while (first_index<=last_index){
    int medium = first_index + (last_index - first_index)/2;
    if (arr[medium]==target){
      return medium;
    }
    if (arr[medium]<target){
      first_index = medium + 1;
    }
    else{
      last_index = medium - 1;
    }
  }
  return -1;     //return -1 if the target not found
}

int main(){
  int arr[] = {2,4,8,16,32,64,128,256};
  int size = sizeof(arr)/sizeof(arr[0]);
  int target = 8;
  int result = binarySearch(arr, 0, size-1, target);
  if (result == 1){
    printf("Taget not found");
  }
  printf("\nTarget is present at index %d\n\n\n", result);
  return 0;
}
